### PR TITLE
Fix some function pointer types to match, i.e. so C++ compiles w/o extra casts.

### PR DESCRIPTION
**Pull request by jaykrell**

### CODE REVIEW

File: mono/metadata/gc-internals.h [hunk 1]
```diff
@@ -140,7 +140,9 @@ MonoGCDescriptor mono_gc_make_descr_for_string (gsize *bitmap, int numbits);
 
 void mono_gc_register_obj_with_weak_fields (void *obj);
 
-void  mono_gc_register_for_finalization (MonoObject *obj, void *user_data);
+typedef void (*MonoFinalizationProc)(void*, void*); // same as GC_finalization_proc
```
Comment (luhenry) [inline mono/metadata/gc-internals.h]: Use `gpointer`.

File: mono/sgen/sgen-gc.h [hunk 2]
```diff
@@ -827,7 +827,10 @@ void sgen_null_link_in_range (int generation, ScanCopyContext ctx, gboolean trac
 void sgen_process_fin_stage_entries (void)
 	MONO_PERMIT (need (sgen_gc_locked));
 gboolean sgen_have_pending_finalizers (void);
-void sgen_object_register_for_finalization (GCObject *obj, void *user_data)
+
+typedef void (*SGenFinalizationProc)(void*, void*); // same as MonoFinalizationProc, GC_finalization_proc
```
Comment (luhenry) [inline mono/sgen/sgen-gc.h]: Use `gpointer`.

Comment (jaykrell) [reply mono/sgen/sgen-gc.h]: The point is to have correct types where it is easy/possible and avoid casts. In this case it is easy. The pointers are always of a precise type and the casts can be removed.

Comment (lambdageek) [reply mono/sgen/sgen-gc.h]: @luhenry you mean `typedef void (*SGenFinalizationProc)(gpointer, gpointer);` right?

Comment (jaykrell) [reply mono/sgen/sgen-gc.h]: Oh, sure, yes, I thought you meant replace the entire pointer with gpointer.

Comment (jaykrell) [reply mono/sgen/sgen-gc.h]: My misunderstanding, sorry, agreed.

**Pull request merged**